ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -117,76 +132,42 @@ delayedProgressReporting before after (Just lspEnv) optProgressStyle = do
-                LSP.WorkDoneProgressCreateParams { _token = u } $ liftIO . signalBarrier b
-            liftIO $ async $ do
-                ready <- waitBarrier b
-                LSP.runLspT lspEnv $ for_ ready $ const $ bracket_ (start u) (stop u) (loop u 0)
+            cancelProgress <- newEmptyMVar
+            LSP.runLspT lspEnv $ do
+                u <- ProgressToken . InR . T.pack . show . hashUnique <$> liftIO newUnique
+                b <- liftIO newBarrier
+                LSP.sendRequest SMethod_WindowWorkDoneProgressCreate
+                    LSP.WorkDoneProgressCreateParams { _token = u } $ liftIO . signalBarrier b
+                async $ do
+                    ready <- liftIO $ waitBarrier b
+                    withProgress "Processing" (Just u) Cancellable $ \update -> loopUntil cancelProgress update 0
+            return cancelProgress
             where
-                start token = LSP.sendNotification SMethod_Progress $
-                    LSP.ProgressParams
-                        { _token = token
-                        , _value = toJSON $ WorkDoneProgressBegin
-                          { _kind = AString @"begin"
-                          ,  _title = "Processing"
-                          , _cancellable = Nothing
-                          , _message = Nothing
-                          , _percentage = Nothing
-                          }
-                        }
-                stop token = LSP.sendNotification SMethod_Progress
-                    LSP.ProgressParams
-                        { _token = token
-                        , _value = toJSON $ WorkDoneProgressEnd
-                          { _kind = AString @"end"
-                           , _message = Nothing
-                          }
-                        }
+                loopUntil m a b = untilMVar m $ loop a b
                 loop _ _ | optProgressStyle == NoProgress =
                     forever $ liftIO $ threadDelay maxBound
-                loop token prevPct = do
+                loop update prevPct = do
                     done <- liftIO $ readTVarIO doneVar
                     todo <- liftIO $ readTVarIO todoVar
                     liftIO $ sleep after
-                    if todo == 0 then loop token 0 else do
-                        let
-                            nextFrac :: Double
+                    if todo == 0 then loop update 0 else do
+                        let nextFrac :: Double
                             nextFrac = fromIntegral done / fromIntegral todo
                             nextPct :: UInt
                             nextPct = floor $ 100 * nextFrac
                         when (nextPct /= prevPct) $
-                          LSP.sendNotification SMethod_Progress $
-                          LSP.ProgressParams
-                              { _token = token
-                              , _value = case optProgressStyle of
```

Comment:
We've lost this progress-style configurability, but honestly, configurability is bad, so better to get rid of it IMO!

Reply:
We should get rid of the option as well, I guess.